services/discovery/client_test.go

```diff
@@ -0,0 +1,141 @@
+package discovery_test
+
+import (
+	"context"
+	"net"
+	"testing"
+
+	"go.viam.com/test"
+	"go.viam.com/utils/rpc"
+
+	viamgrpc "go.viam.com/rdk/grpc"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/discovery"
+	"go.viam.com/rdk/testutils"
+	"go.viam.com/rdk/testutils/inject"
+)
+
+func TestClient(t *testing.T) {
+	logger := logging.NewTestLogger(t)
+	listener1, err := net.Listen("tcp", "localhost:0")
+	test.That(t, err, test.ShouldBeNil)
+	rpcServer, err := rpc.NewServer(logger, rpc.WithUnauthenticated())
+	test.That(t, err, test.ShouldBeNil)
+
+	workingDiscovery := &inject.DiscoveryService{}
+	failingDiscovery := &inject.DiscoveryService{}
+
+	testComponents := []resource.Config{createTestComponent("component-1"), createTestComponent("component-2")}
+
+	workingDiscovery.DiscoverResourcesFunc = func(ctx context.Context, extra map[string]any) ([]resource.Config, error) {
+		return testComponents, nil
+	}
+	failingDiscovery.DiscoverResourcesFunc = func(ctx context.Context, extra map[string]any) ([]resource.Config, error) {
+		return nil, errDiscoverFailed
+	}
+	workingDiscovery.DoFunc = testutils.EchoFunc
+	failingDiscovery.DoFunc = func(
+		ctx context.Context,
+		cmd map[string]interface{},
+	) (
+		map[string]interface{},
+		error,
+	) {
+		return nil, errDoFailed
+	}
```

Comment:
I like this organization better because the working service and failing service functions are grouped together - but if you like yours please feel free to leave as is and resolve this comment. Very minor point here as this is a test.
```suggestion
	testComponents := []resource.Config{createTestComponent("component-1"), createTestComponent("component-2")}
	
	workingDiscovery := &inject.DiscoveryService{}
	workingDiscovery.DiscoverResourcesFunc = func(ctx context.Context, extra map[string]any) ([]resource.Config, error) {
		return testComponents, nil
	}
	workingDiscovery.DoFunc = testutils.EchoFunc
	
	failingDiscovery := &inject.DiscoveryService{}
	failingDiscovery.DiscoverResourcesFunc = func(ctx context.Context, extra map[string]any) ([]resource.Config, error) {
		return nil, errDiscoverFailed
	}
	failingDiscovery.DoFunc = func(
		ctx context.Context,
		cmd map[string]interface{},
	) (
		map[string]interface{},
		error,
	) {
		return nil, errDoFailed
	}
```

Reply:
rearranged the code so the working & failing injects were separated. Note because of the structure of the server tests I couldn't really do the same